### PR TITLE
Implement filling Webforms in HTTP/HTTPS plugin

### PR DESCRIPTION
**Pull request by Lillecarl**
Support for filling forms in HTTP and HTTPS plugins

### CODE REVIEW

**Comment by Lillecarl** [comment]
Thinking about it, one could be lazy and set a fixed amount of ID/Value pairs in the WinForm config, and one for submitting/action. This is probably the easiest most reliable way to do it.

**Comment by jirkapok** [comment]
Please add missing unit tests, fix WebProtocols_CreateControls_ReturnsEmpty test.

**Comment by Lillecarl** [comment]
Considering i don't know how to run the tests, i hope this is what you're looking for :)

**Comment by jirkapok** [comment]
Still two tests are failing. I am going to have a look my self. So we can merge it.

File: Source/Terminals.Common/Connections/KnownConnectionConstants.cs [hunk 1]
```diff
@@ -6,6 +6,8 @@ public class KnownConnectionConstants
 
         public const int HTTPPort = 80;
 
+        public const int HTTPSPort = 443;
```
Comment (jirkapok) [inline Source/Terminals.Common/Connections/KnownConnectionConstants.cs]: There should be no reason why to move it here, until there is another usage than tests. Leave it in https plugin.

File: Source/Terminals.Plugins.Web/HttpsConnectionPlugin.cs [hunk 2]
```diff
@@ -37,12 +41,17 @@ public ProtocolOptions CreateOptions()
 
         public Image GetIcon()
         {
-            return HttpConnectionPlugin.TreeIconHttp;
+            return TreeIconHttps;
```
Comment (jirkapok) [inline Source/Terminals.Plugins.Web/HttpsConnectionPlugin.cs]: Why is the icon moved to this plugin and not referencing the original icon. It looks like a bug?

File: Source/Terminals.Plugins.Web/WebOptions.cs [hunk 3]
```diff
@@ -12,12 +12,26 @@ namespace Terminals.Data
     public class WebOptions : ProtocolOptions, IRelativeUrlProvider
     {
         public string RelativeUrl { get; set; }
+        public string UsernameID { get; set; }
+        public string PasswordID { get; set; }
+        public string OptionalID { get; set; }
+        public string OptionalValue { get; set; }
+        public string SubmitID { get; set; }
+        public bool EnableHTMLAuth { get; set; }
```
Comment (jirkapok) [inline Source/Terminals.Plugins.Web/WebOptions.cs]: Why is the EnableHtmlAuth separated from EnableFormsAuth, what is the difference. I would expect one dropdown to select authentication method. Or is the purpose as fallback authentication?

File: Source/Terminals.Plugins.Web/WebOptionsConverter.cs [hunk 4]
```diff
@@ -1,18 +1,31 @@
...
 
 namespace Terminals.Plugins.Web
 {
-    internal class WebOptionsConverter : IOptionsConverter
+    internal class WebOptionsConverter : OptionsConverterTemplate<WebOptions>, IOptionsConverter
     {
-        public void FromConfigFavorite(OptionsConversionContext context)
+        protected override void FromConfigFavorite(FavoriteConfigurationElement source, WebOptions options)
         {
-            UrlConverter.UpdateFavoriteUrl(context.Favorite, context.ConfigFavorite.Url);
```
Comment (jirkapok) [inline Source/Terminals.Plugins.Web/WebOptionsConverter.cs]: Why are this property conversions removed? The property is still there, so it needs to be converted.

File: Source/Terminals.Plugins.Web/TerminalsWebExport.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        public void ExportOptions(IExportOptionsContext context)
+        {
+            if (context.Favorite.Protocol.Contains("HTTP"))
+                TerminalsWebExport.ExportOptions(ref context);
+        }
+    }
+
+    internal class TerminalsWebExport
+    {
+        public static void ExportOptions(ref IExportOptionsContext context)
```
Comment (jirkapok) [inline Source/Terminals.Plugins.Web/TerminalsWebExport.cs]: Why are you calling the context using ref?

File: Source/Terminals.Plugins.Web/TerminalsWebExport.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    internal class TerminalsHTTPExport : ITerminalsOptionsExport
+    {
+        public void ExportOptions(IExportOptionsContext context)
+        {
+            if (context.Favorite.Protocol.Contains("HTTP"))
+                TerminalsWebExport.ExportOptions(ref context);
+        }
+    }
+
+    internal class TerminalsHTTPSExport : ITerminalsOptionsExport
```
Comment (jirkapok) [inline Source/Terminals.Plugins.Web/TerminalsWebExport.cs]: No need to have to exporters, if they do the same. We may implement one for both protocols. See ConnectionManager.IsProtocolWebBased.

**Comment by jirkapok** [comment]
To fix the tests:
* GetTerminalsOptionsExporters_ReturnsAllExporters: Add newly added exporters class names to expected array.
* AllKnownProtocols_GetFavoriteIcon_ReturnPluginIcon: Fix the reference to HttpsConnectionPlugin.TreeIconHttps for Https test case.

**Review by jirkapok — CHANGES_REQUESTED** [review]

File: Source/Terminals.Plugins.Web/HttpsConnectionPlugin.cs (hunk 2 above)
Comment (Lillecarl) [reply Source/Terminals.Plugins.Web/HttpsConnectionPlugin.cs]: internal static readonly Image TreeIconHttps = Resources.treeIcon_http; I added this to "sync" the two plugins to be as equal as possible, could absolutely be reverted :)

File: Source/Terminals.Plugins.Web/WebOptions.cs (hunk 3 above)
Comment (Lillecarl) [reply Source/Terminals.Plugins.Web/WebOptions.cs]: Added both to be able to keep both enabled by default, to ease for users (Not having to chose by default).

File: Source/Terminals.Plugins.Web/TerminalsWebExport.cs (hunk 6 above)
Comment (Lillecarl) [reply Source/Terminals.Plugins.Web/TerminalsWebExport.cs]: Correcting this :)

File: Source/Terminals.Plugins.Web/WebOptionsConverter.cs (hunk 4 above)
Comment (Lillecarl) [reply Source/Terminals.Plugins.Web/WebOptionsConverter.cs]: I'll look into it :)

**Comment by Lillecarl** [comment]
Updated to try and fit your comments. It might still not be perfect for your style though. We'll see :)

**Comment by jirkapok** [comment]
Please add missing tests to cover what is not automatically tested. Otherwise i will fix the issues you dint comment.

**Comment by Lillecarl** [comment]
If there's anything you want fixed, feel free to push to my branch too, you should have write access to the pullrequested branches :)

My reasoning behind instancing the icon twice and moving the https port definition is only to keep consistency within the codebase.

Hope to see this if not merged, implemented :) It's a killer feature! :)

**Comment by Lillecarl** [comment]
If you want, we can rebase this branch on master and rewrite the history too :) Contributed to a project awhile ago where a pretty gitlog was important.

**Comment by Lillecarl** [comment]
How are we doing on this? I'm not sure which changes i've missed, please rereview this.

**Comment by jirkapok** [comment]
Go through all comments for each file, mainly:
* Add tests fro the newly added values
* Implement upgrade from previous version

**Comment by Lillecarl** [comment]
@jirkapok It's the wrong way around, but i'd love it if you helped me cleanup the last bits as you'd like them. I think by GitHub design you've got write access to the pullrequest branch.

The tests i think i implemented, i'm not sure what's missing. And upgrades i know not how to do.

Best Regards
Carl :)

**Comment by jirkapok** [comment]
Based on project resurection, i will cherry pick these changes and implement them in 46-putty branch.

**Comment by Lillecarl** [comment]
I'm glad to hear, I don't have any Windows machines anymore so I can't assist, but I suspekt you've got this 😄

**Pull request closed without merging**